### PR TITLE
Dask Dashboard

### PR DESCRIPTION
**Pull request by jcrist**
A dashboard for visualizing dask execution using bokeh. There are so many more things I want to add to this, but here's the initial code. Based on #356. Note that this requires you have bokeh installed, and `bokeh-server` running.

Example use:
```python
>>> import dask.array as da
>>> from dask.array.core import get
>>> import numpy as np
>>> from dask.dashboard import Dashboard

# Create a dashboard, with a 1 second delay between scheduler ticks
>>> dash = Dashboard(get, pause=1)

>>> a = da.from_array(np.arange(36).reshape((6, 6)), chunks=(3, 3))
>>> res = a.T.dot(a).mean() + a.var()

# Show the dask in the dashboard
>>> dash.visualize(res.dask)
    
# Compute result, displaying the progress in the visualizer
>>> res.compute(get=dash.get)
```

demo

### CODE REVIEW

**Comment by cowlicks** [comment]
That is a pretty slick gif! It would be cool to have the output in an IPython notebook.

File: dask/dashboard/vis_app.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    def __init__(self, vis):
+        super(_Handler, self).__init__()
+        self.queue = vis._queue
+        self.update_dask = vis.update_dask
+        self.daemon = True
+
+    def run(self):
+        while True:
+            task = self.queue.get()
+            self.update_dask(*task)
```
Comment (mrocklin) [inline dask/dashboard/vis_app.py]: I wonder if we can get someone familiar with bokeh to review this part.

cc @bryevdv, @fpliger, or @chdoig

File: dask/dashboard/to_pydot.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                dep_name = name(dep)
+                if dep_name not in node_lk:
+                    node_lk[dep_name] = node = Node(dep_name, label=str(dep),
+                                                    shape='box')
+                    g.add_node(node)
+                g.add_edge(Edge(dep_name, func_name, arrowhead='none'))
+        elif ishashable(v) and v in d:
+            g.add_edge(Edge(name(v), k_name, arrowhead='none'))
+        g.set_rankdir('BT')
+    return g
```
Comment (mrocklin) [inline dask/dashboard/to_pydot.py]: Should this replace the `dot.py` code currently in master?

File: dask/dashboard/graphparse.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            edges.append(Edge._from_parse(data))
+        elif line[0] == 'stop':
+            break
+    return Graph(scale, width, height, nodes, edges)
+
+
+def dask_to_graph(dsk):
+    p = to_pydot(dsk)
+
+    out, ret = subprocess.Popen(['dot', '-Tplain'], stdin=subprocess.PIPE,
```
Comment (mrocklin) [inline dask/dashboard/graphparse.py]: I had no idea that `plain` existed as an output type.

File: dask/async.py [hunk 4]
```diff
@@ -414,29 +429,27 @@ def get_async(apply_async, num_workers, dsk, result, cache=None,
 
     state = start_state_from_dask(dsk, cache=cache)
 
-    tick = [0]
-
     if state['waiting'] and not state['ready']:
         raise ValueError("Found no accessible jobs in dask")
 
+    if not execute_cm:
+        execute_cm = default_cm
```
Comment (mrocklin) [inline dask/async.py]: Can this be handled as a kwarg?

File: dask/dashboard/to_pydot.py (hunk 2 above)
Comment (jcrist) [inline dask/dashboard/to_pydot.py]: It should, just need to add in the kwargs that you had there. Unless those aren't useful anymore. Note that this doesn't depend on networkx anymore.

File: dask/async.py (hunk 4 above)
Comment (jcrist) [inline dask/async.py]: It could, I just thought `None` made more sense to be user facing than an opaque function. A default would be fine too.

Comment (mrocklin) [inline dask/async.py]: No strong opinion

**Comment by jcrist** [comment]
Just a reminder to myself to rebase this on top of the callbacks PR once that's merged.

File: dask/dashboard/vis_app.py (hunk 1 above)
Comment (bryevdv) [inline dask/dashboard/vis_app.py]: As is it you are mixing the low level "models" API with the mid level "plotting" API. That's not necessarily a bad thing but I also think here it is necessary, you you could just use e.g. `self.plot.text(...)` and `self.plot.circle(...)` and then you would not need all the `add_glyph` calls.

Comment (jcrist) [inline dask/dashboard/vis_app.py]: This is leftover from before, when I was creating the whole figure once, and then changing the data. But then I found I couldn't link fontsize to data, and the reset button required a new plot, so I had to make a new plot for each dask.

Out of curiosity, is there a performance difference between:
```
glyphs = make_the_glyphs(...)    # Run only once
for stuff in loop:
    fig = figure(...)
    fig.add_glyphs(glyphs)
```

and 

```
for stuff in loop:
    fig = figure()
    fig.glyph(...)
    fig.other_glyph(...)
```

Comment (bryevdv) [inline dask/dashboard/vis_app.py]: I wouldn't expect an appreciable one, if any. I should mention that re-using glyphs in this way is probably not something that has been tested rigorously, if at all.

Comment (fpliger) [inline dask/dashboard/vis_app.py]: I'd recommend to not re-use the glyphs this way... better to just recreate them. I've run into problems reusing glyphs recently but didn't have time to dig into that and formalize/reproduce the issue.

**Comment by mrocklin** [comment]
I think that something like this should go in to `dask.diagnostics`.  It's fairly useful for debugging.  I wouldn't mind seeing another version with the same output style as the current Profiler, which I think is useful for monitoring.

**Pull request closed without merging**